// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// The following only applies to changes made to this file as part of YugaByte development.
//
// Portions Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//
// Protobufs which are common throughout YB.
//
// This file may contain protobufs which are persisted on disk
// as well as sent on the wire. If a particular protobuf is only
// used as part of the client-server wire protocol, it should go
// in common/wire_protocol.proto instead. If it is only used within
// the server(s), it should go in server/metadata.proto,
// etc, as appropriate.
syntax = "proto2";

package yb;

option java_package = "org.yb";

// Client type.
enum QLClient {
  YQL_CLIENT_CQL = 1;
  YQL_CLIENT_PGSQL = 2;
  YQL_CLIENT_REDIS = 3;
}

// Client type.
enum YQLDatabase {
  YQL_DATABASE_UNKNOWN = 1;
  YQL_DATABASE_CQL = 2;
  YQL_DATABASE_PGSQL = 3; // Pgsql database
  YQL_DATABASE_REDIS = 4;
}

// To ensure compatibility between release versions, the numeric values of these datatypes cannot
// be changed once the types are implemented and released.
//
// Make sure this is in sync with YBCPgDataType in ybc_pg_typedefs.h.
enum DataType {
  UNKNOWN_DATA = 999;
  NULL_VALUE_TYPE = 0;
  INT8 = 1;
  INT16 = 2;
  INT32 = 3;
  INT64 = 4;
  STRING = 5;
  BOOL = 6;
  FLOAT = 7;
  DOUBLE = 8;
  BINARY = 9;
  TIMESTAMP = 10;
  DECIMAL = 11;
  VARINT = 12;
  INET = 13;
  LIST = 14;
  MAP = 15;
  SET = 16;
  UUID = 17;
  TIMEUUID = 18;
  TUPLE = 19;  // TUPLE is not yet fully implemented, but it is a CQL type.
  TYPEARGS = 20;
  USER_DEFINED_TYPE = 21;
  FROZEN = 22;
  DATE = 23;
  TIME = 24;
  JSONB = 25;

  // All unsigned datatypes will be removed from QL because databases do not have these types.
  UINT8 = 100;
  UINT16 = 101;
  UINT32 = 102;
  UINT64 = 103;
}

// This must exactly match YBTableType in client.h.
// We have static_assert's in tablet-test.cc to verify this.
enum TableType {
  option allow_alias = true;

  YQL_TABLE_TYPE = 2;
  REDIS_TABLE_TYPE = 3;
  PGSQL_TABLE_TYPE = 4;
  TRANSACTION_STATUS_TABLE_TYPE = 5;

  DEFAULT_TABLE_TYPE = 2;
}

enum YBConsistencyLevel {
  // This consistency level provides Linearizability guarantees and is the default for our system.
  STRONG = 1;

  // Consistent-prefix consistency means that we always see consistent snapshot of the database in
  // a well-defined order. If operations A, B and C take place, we will either see A, AB, or ABC.
  // Note that reads might still go back in time since we might see ABC on one replica and AB on
  // another.
  CONSISTENT_PREFIX = 2;

  // For cross-shard transactions only: user-enforced consistency level means it is the user's
  // responsibility to enforce consistency across shards or tables/indexes.
  USER_ENFORCED = 3;
}

// Used for Cassandra Roles and Permissions
enum ResourceType {
  ALL_KEYSPACES = 1;
  KEYSPACE = 2;
  TABLE = 3;
  ALL_ROLES = 4;
  ROLE = 5;
}

// Suffixed with PERMISSION, because Google does not allow same enum name CREATE
enum PermissionType {
  ALTER_PERMISSION = 0;
  CREATE_PERMISSION = 1;
  DROP_PERMISSION = 2;
  SELECT_PERMISSION = 3;
  MODIFY_PERMISSION = 4;
  AUTHORIZE_PERMISSION = 5;
  DESCRIBE_PERMISSION = 6;
  ALL_PERMISSION = 999999999;
}

// The type used in column schemas, which may have type parameters (i.e. for collections)
// If types have parameters, they are stored in the params field. Otherwise params is empty.
// e.g. (using lower case of QLTypes and upper case for DataType):
//   string                     --> {main = STRING, params = []}
//   set<int>                   --> {main = SET, params = [int]}
//   map<int, string>           --> {main = MAP, params = [int, string]}
// later (not implemented yet):
//   tuple<int, string, double> --> {main = TUPLE, params = [int, string, double]}
// User-Defined types are identified by keyspace_name and type_name and their fields are stored in
// field_names (for the names) and params (for the types)
// e.g. given a declared type "ks.employee (firstname : string, lastname : string, ssn : bigint)":
// ks.employee                  --> { main = USER_DEFINED_TYPE,
//                                    params = [string, string, bigint],
//                                    keyspace_name = "ks",
//                                    type_name = "employee",
//                                    field_names = ["firstname", "lastname", "ssn"]}

message QLTypePB {
  optional DataType main = 1;
  repeated QLTypePB params = 2;

  // Fields for user-defined types
  message UDTypeInfo {
    optional string keyspace_name = 1;
    optional string name = 2;
    optional string id = 3;
    repeated string field_names = 4;
  }
  optional UDTypeInfo udtype_info = 3;
}

// TODO: Differentiate between the schema attributes
// that are only relevant to the server (e.g.,
// encoding and compression) and those that also
// matter to the client.
message ColumnSchemaPB {
  optional uint32 id = 1;
  required string name = 2;
  required QLTypePB type = 3;
  optional bool is_key = 4 [default = false];
  optional bool is_hash_key = 5 [default = false];
  optional bool is_nullable = 6 [default = false];
  optional bool is_static = 7 [default = false];
  optional bool is_counter = 8 [default = false];
  optional int32 order = 12 [default = 0];
  optional uint32 sorting_type = 9 [default = 0];
  // Reserved for deprecated read_default_value and write_default_value fields.
  reserved 10, 11;
  // The JSON attribute was mistakenly placed here, but it is not used.
  // To be safe, it's left here for now.
  // JSON attribute (for c->'a'->>'b' case).
  repeated QLJsonOperationPB OBSOLETE_json_operations = 13;
}

message TablePropertiesPB {
  optional uint64 default_time_to_live = 1;
  optional bool contain_counters = 2;
  optional bool is_transactional = 3 [default = false];

  // The table id of the table that this table is co-partitioned with.
  optional bytes copartition_table_id = 4;

  // For index table only: consistency with respect to the indexed table.
  optional YBConsistencyLevel consistency_level = 5 [ default = STRONG ];

  // For CREATE INDEX, we use mangled name for its column, but it hasn't been that way.
  // Due to coding history for INDEX support, we'll have the following scenarios.
  // - UserTable: User-defined names are used for its columns.
  // - IndexTable: Mangled names are used. We treat all INDEX column definition as expression of
  //   column. For example INDEX on TAB(a) would be defined as IndexTable( ref_of(a) ).
  // - Older IndexTable: User-defined name is used, and their definition in existing database
  //   remains the same during upgrade.
  //
  // The field "use_mangled_column_name" helps indicating whether a table is using mangled_name.

  optional bool use_mangled_column_name =  6 [ default = false ];
  optional int32 num_tablets = 7 [ default = 0 ];
  optional bool is_ysql_catalog_table = 8 [ default = false ];
}

message SchemaPB {
  repeated ColumnSchemaPB columns = 1;
  optional TablePropertiesPB table_properties = 2;
}

// This message contains the metadata of a secondary index of a table.
// It maps the index::columns to the expressions of table::columns.
//
// Notes on where to find metadata of an INDEX.
// - Each INDEX is represented by IndexInfo and IndexTable.
//     IndexInfoPB contains the definitions of the INDEX.
//     IndexTable contains duplicate user-data for quick access.
// - The IndexInfoPB is stored in the Catalog::Table being indexed.
// - The TablePropertiesPB is kept in Catalog::IndexTable.
//   Internally, Catalog::IndexTable is just the same as any Catalog::Table.
message IndexInfoPB {
  optional bytes table_id = 1; // Index table id.
  optional bytes indexed_table_id = 8; // Indexed table id.
  optional uint32 version = 2 [ default = 0]; // Index table's schema version.
  optional bool is_local = 3 [ default = false ];  // Whether the index is a local index
  optional bool is_unique = 7 [ default = false ]; // Whether the index is a unique index

  // Index column mapping.
  // "colexpr" is used to compute the value of this column in an INDEX.
  // - When a table is indexed by expressions, we create internal/hidden columns to store the index
  //   value, and "value_expr" specifies the indexing expression.
  // - As of 07/2019, only QLJsonColumnOperationsPB is allowed for "colexpr".
  // - In the current index design & implementation, expression can only reference ONE column.
  //
  // Example:
  //   Example for scalar index
  //     TABLE (a, b, c)
  //     INDEX (c) -> INDEX is a table whose column 'c' is referencing TABLE(c)
  //     colexpr = ref to "c" column_id.
  //   Example for JSON index
  //     TABLE (a, b, j)
  //     INDEX (j->>'b') -> INDEX is a table whose column 'j->>b' is referencing to TABLE(j)
  //     colexpr = j->'b'
  message IndexColumnPB {
    optional uint32 column_id = 1;         // Column id in the index table.
    optional string column_name = 3;       // Generated column name in the index table.
    optional uint32 indexed_column_id = 2; // Corresponding column id in indexed table.
    optional QLExpressionPB colexpr = 4;   // Column value in INDEX.
  }
  repeated IndexColumnPB columns = 4;      // Indexed and covering columns.
  optional uint32 hash_column_count = 5;   // Number of hash columns in the index.
  optional uint32 range_column_count = 6;  // Number of range columns in the index.

  repeated uint32 indexed_hash_column_ids = 9;   // Hash column ids in the indexed table.
  repeated uint32 indexed_range_column_ids = 10; // Range column ids in the indexed table.

  // The mangled-name flag is kept on both IndexInfo and IndexTable as the same mangled-name is
  // used in both IndexInfo and IndexTable columns.
  optional bool use_mangled_column_name = 11 [ default = false ];  // Newer index has mangled name.
}

message HostPortPB {
  required string host = 1;
  required uint32 port = 2;
}

// The possible order modes for clients.
// Clients specify these in new scan requests.
// Ordered scans are fault-tolerant, and can be retried elsewhere in the case
// of tablet server failure. However, ordered scans impose additional overhead
// since the tablet server needs to sort the result rows.
enum OrderMode {
  UNKNOWN_ORDER_MODE = 0;
  // This is the default order mode.
  UNORDERED = 1;
  ORDERED = 2;
}

// The serialized format of a YB table partition schema.
message PartitionSchemaPB {

  // A column identifier for partition schemas. In general, the name will be
  // used when a client creates the table since column IDs are assigned by the
  // master. All other uses of partition schemas will use the numeric column ID.
  message ColumnIdentifierPB {
    oneof identifier {
      int32 id = 1;
      string name = 2;
    }
  }

  message RangeSchemaPB {
    // Column identifiers of columns included in the range. All columns must be
    // a component of the primary key.
    repeated ColumnIdentifierPB columns = 1;
  }

  message HashBucketSchemaPB {
    // Column identifiers of columns included in the hash. Every column must be
    // a component of the primary key.
    repeated ColumnIdentifierPB columns = 1;

    // Number of buckets into which columns will be hashed. Must be at least 2.
    required int32 num_buckets = 2;

    // Seed value for hash calculation. Administrators may set a seed value
    // on a per-table basis in order to randomize the mapping of rows to
    // buckets. Setting a seed provides some amount of protection against denial
    // of service attacks when the hash bucket columns contain user provided
    // input.
    optional uint32 seed = 3;

    enum HashAlgorithm {
      UNKNOWN = 0;
      MURMUR_HASH_2 = 1;
    }

    // The hash algorithm to use for calculating the hash bucket.
    optional HashAlgorithm hash_algorithm = 4;
  }

  repeated HashBucketSchemaPB hash_bucket_schemas = 1;
  optional RangeSchemaPB range_schema = 2;

  enum HashSchema {
    MULTI_COLUMN_HASH_SCHEMA = 1;
    REDIS_HASH_SCHEMA = 2;
    PGSQL_HASH_SCHEMA = 3;
  }

  optional HashSchema hash_schema = 3;
}

// The serialized format of a YB table partition.
message PartitionPB {
  // The hash buckets of the partition. The number of hash buckets must match
  // the number of hash bucket components in the partition's schema.
  repeated int32 hash_buckets = 1 [packed = true];
  // The encoded start partition key (inclusive).
  optional bytes partition_key_start = 2;
  // The encoded end partition key (exclusive).
  optional bytes partition_key_end = 3;
}

enum IsolationLevel {
  NON_TRANSACTIONAL = 0;
  SNAPSHOT_ISOLATION = 1;
  SERIALIZABLE_ISOLATION = 2;
}

// This enum matches enum RowMarkType defined in src/include/nodes/plannodes.h.
// The exception is ROW_MARK_ABSENT, which signifies the absence of a row mark.
enum RowMarkType
{
  // Obtain exclusive tuple lock.
  ROW_MARK_EXCLUSIVE = 0;

  // Obtain no-key exclusive tuple lock.
  ROW_MARK_NOKEYEXCLUSIVE = 1;

  // Obtain shared tuple lock.
  ROW_MARK_SHARE = 2;

  // Obtain keyshare tuple lock.
  ROW_MARK_KEYSHARE = 3;

  // Not supported. Used for postgres compatibility.
  ROW_MARK_REFERENCE = 4;

  // Not supported. Used for postgres compatibility.
  ROW_MARK_COPY = 5;

  // Obtain no tuple lock (this should never sent be on the wire).  The value
  // should be high for convenient comparisons with the other row lock types.
  ROW_MARK_ABSENT = 15;
}

enum TransactionStatus {
  CREATED = 1;
  PENDING = 2;

  COMMITTED = 4;
  APPLIED_IN_ALL_INVOLVED_TABLETS = 7;

  ABORTED = 8;

  // All following entries are not used in RAFT, but as events between status tablet and involved
  // tablets:
  APPLYING = 20;
  APPLIED_IN_ONE_OF_INVOLVED_TABLETS = 21;
  CLEANUP = 22;
}

message TransactionMetadataPB {
  optional bytes transaction_id = 1; // 16 byte uuid
  optional IsolationLevel isolation = 2 [default=NON_TRANSACTIONAL];
  optional string status_tablet = 3;
  optional fixed64 priority = 4;

  optional fixed64 start_hybrid_time = 5;

  // Stores time when metadata was written to provisional records RocksDB on a
  // participating tablet. So it could be used for cleanup.
  optional fixed64 metadata_write_time = 6;
}

// See ReadHybridTime for explation of this message.
message ReadHybridTimePB {
  optional fixed64 read_ht = 1;

  optional fixed64 local_limit_ht = 2;

  optional fixed64 global_limit_ht = 3;

  optional fixed64 in_txn_limit_ht = 4;
}

// For clarification of field meaning see comments of appropriate fields in YBTransaction::Impl
message ChildTransactionDataPB {
  optional TransactionMetadataPB metadata = 1;
  optional ReadHybridTimePB read_time = 2;
  map<string, fixed64> local_limits = 3;
}

enum InvolvedTabletMetadataState {
  MISSING = 0;
  EXIST = 1;
  MAY_EXIST = 2;
}

message TransactionInvolvedTabletPB {
  optional bytes tablet_id = 1;
  optional InvolvedTabletMetadataState metadata_state = 2;
}

// For clarification of field meaning see comments of appropriate fields in YBTransaction::Impl
message ChildTransactionResultPB {
  repeated TransactionInvolvedTabletPB tablets = 1;
  optional fixed64 restart_read_ht = 2;
  map<string, fixed64> read_restarts = 3;
  optional ReadHybridTimePB used_read_time = 4;
}

message DeletedColumnPB {
  // Required. Column Id.
  optional int32 column_id = 1;

  // Required. Hybrid Time when column was deleted on this tablet.
  optional fixed64 deleted_hybrid_time = 2;
}

message CloudInfoPB {
  optional string placement_cloud = 1;
  optional string placement_region = 2;
  optional string placement_zone = 3;
}

//--------------------------------------------------------------------------------------------------
// YQL support.
// The following section is common for all query language.
//--------------------------------------------------------------------------------------------------

// A QL value
message QLValuePB {
  // Note: an absent value means NULL
  oneof value {
    int32 int8_value = 1;  // Note: min int size in protobuf is int32
    int32 int16_value = 2;
    int32 int32_value = 3;
    int64 int64_value = 4;
    float float_value = 5;
    double double_value = 6;
    string string_value = 7;
    bool bool_value = 8;
    int64 timestamp_value = 9;
    bytes binary_value = 10;
    // raw bytes for inet address in network byte order.
    bytes inetaddress_value = 11;
    QLMapValuePB map_value = 12;
    QLSeqValuePB set_value = 13;
    QLSeqValuePB list_value = 14;
    bytes decimal_value = 15;
    bytes varint_value = 16;
    QLSeqValuePB frozen_value = 17;
    // raw bytes for uuid value.
    bytes uuid_value = 18;
    // raw bytes for timeuuid value.
    bytes timeuuid_value = 19;
    bytes jsonb_value = 20;
    // Number of days where 2^31 corresponds to 1970-01-01
    // (see DateTime::DateFromString)
    uint32 date_value = 21;
    // Number of nano-seconds, from 0 to (24 * 60 * 60 * 1,000,000,000 - 1)
    // (see DateTime::kMaxTime)
    int64 time_value = 22;
    uint32 uint32_value = 23;
    uint64 uint64_value = 24;
  }
}

//--------------------------------------------------------------------------------------------------
// CQL support.
// This is an internal protocol for communicating QL operations from a YB client to a tserver.
//
// Reference:
// Cassandra SQL doc - https://docs.datastax.com/en/cql/3.3/cql/cql_reference/cqlCommandsTOC.html
// Usage notes - https://docs.google.com/document/d/1q9i7a9X5rDdTC0q8hNAqo9i9NyDh0n_Fhm5vdzmtBEw
//--------------------------------------------------------------------------------------------------

// Sequence of values used to represent Lists and Sets (and later Tuples)
message QLSeqValuePB {
  repeated QLValuePB elems = 1;
}

message QLMapValuePB {
  repeated QLValuePB keys = 1;
  repeated QLValuePB values = 2;
}

// Reference to a subcolumn, e.g. m['x'] or l[2]['x']
message QLSubscriptedColPB {
  optional int32 column_id = 1;
  repeated QLExpressionPB subscript_args = 2;
}

// Represents operations applied to a json column.
message QLJsonColumnOperationsPB {
  optional int32 column_id = 1;
  repeated QLJsonOperationPB json_operations = 2;
}

enum JsonOperatorPB {
    JSON_OBJECT = 0; // The -> operator applied to a column.
    JSON_TEXT = 1; // The ->> operator applied to a column.
}

message QLJsonOperationPB {
    required JsonOperatorPB json_operator = 1;
    required QLExpressionPB operand = 2;
}

// An expression in a WHERE condition
message QLExpressionPB {
  oneof expr {
    QLValuePB value = 1;
    int32 column_id = 2;
    QLSubscriptedColPB subscripted_col = 3;
    int32 bind_id = 4;                      // Bind variable index.
    QLConditionPB condition = 5;            // This should be replaced with builtin operator.
    QLBCallPB bfcall = 6;                   // Regular builtin calls.
    QLBCallPB tscall = 7;                   // Tablet server builtin calls.
    QLBCallPB bocall = 8;                   // Builtin operator calls.
    QLJsonColumnOperationsPB json_column = 9; // Json column operators.
  }
}

// Expression operators.
enum QLOperator {
  QL_OP_NOOP = 0;

  // Logic operators that take one operand.
  QL_OP_NOT = 1;
  QL_OP_IS_TRUE = 2;
  QL_OP_IS_FALSE = 3;

  // Logic operators that take two or more operands.
  QL_OP_AND = 4;
  QL_OP_OR = 5;

  // Relation operators that take one operand.
  QL_OP_IS_NULL = 6;
  QL_OP_IS_NOT_NULL = 7;

  // Relation operators that take two operands.
  QL_OP_EQUAL = 8;
  QL_OP_LESS_THAN = 9;
  QL_OP_LESS_THAN_EQUAL = 10;
  QL_OP_GREATER_THAN = 11;
  QL_OP_GREATER_THAN_EQUAL = 12;
  QL_OP_NOT_EQUAL = 13;

  QL_OP_LIKE = 14;
  QL_OP_NOT_LIKE = 15;
  QL_OP_IN = 16;
  QL_OP_NOT_IN = 17;

  // Relation operators that take three operands.
  QL_OP_BETWEEN = 18;
  QL_OP_NOT_BETWEEN = 19;

  // Operators that take no operand. For use in "if" clause only currently.
  QL_OP_EXISTS = 20;     // IF EXISTS
  QL_OP_NOT_EXISTS = 21; // IF NOT EXISTS
}

// A logical condition that evaluates to true/false. Used in the WHERE clause.
message QLConditionPB {
  optional QLOperator op = 1;
  repeated QLExpressionPB operands = 2;
}

// Builtin call expression. There are 3 different calls.
// - Builtin operators such as '>', '<', '=', ...
//   These operators can be executed anywhere.
// - Builtin functions such as Now().
//   These functions can be executed anywhere.
// - Server builtin functions.
//   Only tablet servers can execute these functions.
//
// TODO(neil) Regular builtin operators. This message can be executed anywhere.
// - This is more efficient than builtin call as it avoids most overheads of calling builtin lib.
// - Merge the current condition operator execution with this.
// - To optimize certain operation (such as +), replace it builtin function with builtin op.
message QLBCallPB {
  optional int32 opcode = 1;
  repeated QLExpressionPB operands = 2;
}


//--------------------------------------------------------------------------------------------------
// EXPLAIN PLAN
//
// Explain Plan Protobufs for 4 main DDL statements.
// Output examples at https://github.com/YugaByte/yugabyte-db/issues/861#issuecomment-469350077
// TODO: Change from just explain plan strings to proper Query Plan with enums.
// - Cqlsh output is right-justified so 'output_width' is used to right-pad the output so that it
//   appears left-justified like PostgreSQL explain plan.
// - 'key_conditions' is the set of columns of the where clause that are part of the primary key.
// - 'filter' is the set of the rest of the where clause columns are that not in the primary key.
//--------------------------------------------------------------------------------------------------

message ExplainPlanPB {
  oneof plan {
    SelectPlanPB select_plan = 1;
    InsertPlanPB insert_plan = 2;
    UpdatePlanPB update_plan = 3;
    DeletePlanPB delete_plan = 4;
  }
}

message SelectPlanPB {
  optional string select_type = 1;
  // Overarching information like "Aggregate" or "Limit"
  optional string aggregate = 2;
  optional string key_conditions = 3;
  optional string filter = 4;
  optional int32 output_width = 5;
}

message InsertPlanPB {
  optional string insert_type = 1;
  optional int32 output_width = 2;
}

message UpdatePlanPB {
  optional string update_type = 1;
  // For update, scan_type is currently only primary key lookup.
  optional string scan_type = 2;
  optional string key_conditions = 3;
  optional int32 output_width = 4;
}

message DeletePlanPB {
  optional string delete_type = 1;
  // For delete, scan_type is currently either primary key lookup or range scan.
  optional string scan_type = 2;
  optional string key_conditions = 3;
  optional string filter = 4;
  optional int32 output_width = 5;
}
